Support mixing LVI mitigation options (#4715)
openenclave/openenclave: samples/helloworld/README.md

- All the enclave code is compiled with the mitigation.
- All the enclave code is linked against the mitigated version of Open Enclave libraries.

> > If you wish to enable LVI mitigation, it is preferred that you use the same LVI mitigation method as the one used to build OpenEnclave. Mixing them, for instance, building OE with ControlFlow-GNU and your app with ControlFlow-Clang would work, and vice versa, but will require additional configuration at compile time (see [#Mixing-LVI-Mitigation](#Mixing-LVI-Mitigation))

Reviewer: ```suggestion
> If you wish to enable LVI mitigation, it is preferred that you use the same LVI mitigation method as the one used to build Open Enclave. Mixing them, for instance, building Open Enclave with ControlFlow-GNU and your app with ControlFlow-Clang would work, and vice versa, but will require additional configuration at compile time (see [#Mixing-LVI-Mitigation](#Mixing-LVI-Mitigation))
```

ninja run
```

### Mixing LVI Mitigation

> You can use different LVI mitigation method to build your app than OpenEnclave, however some additional cmake configuration is required. See below table for all possible combination.

Reviewer: ```suggestion
You can use a different LVI mitigation method to build your app than Open Enclave, however additional cmake configuration may be required. See below table for all possible combination.
```

From your table there are no additional requirements when building OE with Clang and application with GNU so it is not required for all cases.

> You can use a different LVI mitigation method to build your app than Open Enclave, however additional cmake configuration may be required. See below table for all possible combination.

Reviewer: ```suggestion
You can use a different LVI mitigation method to build your app than Open Enclave, however additional cmake configuration may be required. See the table below for all possible combinations.
```

| OE    | client | Example cmake command                                                                                   | Additional Requirement                                                                                                   |

> |-------|-----------|---------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------|

Reviewer 2: what happens if someone specifies _DLVI_MITIGATION=ControlFlow? Can you default to ControlFlow_Gnu.
Can you also include a note in bold? OE release packages are built with Controlflow Gnu. 
OE built with ControlFlow_Clang is only available when contributors build from source

Author: Changed doc

| GNU   | GNU     | `cmake .. -DLVI_MITIGATION=ControlFlow-GNU -DLVI_MITIGATION_BINDIR=/path/to/lvi_mitigation_bin`           |                                                                                                                          |

> | GNU   | Clang       | `cmake .. -DLVI_MITIGATION=ControlFlow-Clang -DLVI_MITIGATION_BINDIR=/path/to/lvi_mitigation_bin`         | Must pass `OE_USE_CONTROLFLOW_GNU` to `apply_lvi_mitigation`. Example: `apply_lvi_mitigation(enclave, OE_USE_CONTROLFLOW_GNU)` |

Reviewer: ```suggestion
| GNU   | Clang       | `cmake .. -DLVI_MITIGATION=ControlFlow-Clang -DLVI_MITIGATION_BINDIR=/path/to/lvi_mitigation_bin`         | Must pass `OE_USED_CONTROLFLOW_GNU` to `apply_lvi_mitigation`. Example: `apply_lvi_mitigation(enclave, OE_USED_CONTROLFLOW_GNU)` |
```

> | GNU   | Clang       | `cmake .. -DLVI_MITIGATION=ControlFlow-Clang -DLVI_MITIGATION_BINDIR=/path/to/lvi_mitigation_bin`         | Must pass `OE_USED_CONTROLFLOW_GNU` to `apply_lvi_mitigation`. Example: `apply_lvi_mitigation(enclave, OE_USED_CONTROLFLOW_GNU)` |

Reviewer 2: Since we always use OE_USED_CONTROLFLOW_GNU, can we make it the default option. We are not releasing packages with controlflow_clang.
I would remove this table and make the user experience - OE built with controlflow_gnu, in apply_lvi_mitigation, assume that controlflow_gnu was used, unless otherwise specified.

Author: Yes, that sounds more user-friendly. Will do

Author: The commit 08f3fb51d2c444d94098223c3d1a708d61b79b8e address this comment.

A few changes:
1. now we have 2 copies of `apply_lvi_mitigation` that are largely the same, one for building Open Enclave, which doesn't need optional arg to change its behavior, the other provided specifically to clients to build their project, which accepts optional arg
2. For the `apply_lvi_mitigation` used by the client, it defaults to assume OE is built with GNU, and usage remains the same. If OE and client project are both built by clang, then need to pass in `NO_CUSTOM_TOOLCHAIN`

Author: Now the command example looks like this
| OE    | client | Supported | Example cmake command                                                                                   | Additional Requirement                                                                                                   |
|-------|--------|-----------|---------------------------------------------------------------------------------------------------------|--------------------------------------------------------------------------------------------------------------------------|
| GNU   | GNU    | Yes       | `cmake .. -DLVI_MITIGATION=ControlFlow-GNU -DLVI_MITIGATION_BINDIR=/path/to/lvi_mitigation_bin`           |                                                                                                                          |
| GNU   | Clang  | Yes       | `cmake .. -DLVI_MITIGATION=ControlFlow-Clang -DLVI_MITIGATION_BINDIR=/path/to/lvi_mitigation_bin`         |  |
| Clang | GNU    | Yes       | `cmake .. -DLVI_MITIGATION=ControlFlow-GNU -DLVI_MITIGATION_BINDIR=/path/to/lvi_mitigation_bin`           |                                                                                                                          |
| Clang | Clang  | Yes       | `cmake .. -DLVI_MITIGATION=ControlFlow-Clang -DCMAKE_C_COMPILER=clang-11 -DCMAKE_CXX_COMPILER=clang++-11` |      Must pass `NO_CUSTOM_TOOLCHAIN` to `apply_lvi_mitigation`. Example: `apply_lvi_mitigation(enclave, NO_CUSTOM_TOOLCHAIN)`    |

| Clang | GNU         | `cmake .. -DLVI_MITIGATION=ControlFlow-GNU -DLVI_MITIGATION_BINDIR=/path/to/lvi_mitigation_bin`           |                                                                                                                          |

> | Clang | Clang        | `cmake .. -DLVI_MITIGATION=ControlFlow-Clang -DCMAKE_C_COMPILER=clang-11 -DCMAKE_CXX_COMPILER=clang++-11` |      Must **NOT** pass `OE_USE_CONTROLFLOW_GNU` to `apply_lvi_mitigation`.    |

Reviewer: ```suggestion
| Clang | Clang        | `cmake .. -DLVI_MITIGATION=ControlFlow-Clang -DCMAKE_C_COMPILER=clang-11 -DCMAKE_CXX_COMPILER=clang++-11` |      Must **NOT** pass `OE_USED_CONTROLFLOW_GNU` to `apply_lvi_mitigation`.    |
```
